dart/common/SpecializedAddonManager.h

```diff
@@ -0,0 +1,195 @@
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_SPECIALIZEDADDONMANAGER_H_
+#define DART_COMMON_SPECIALIZEDADDONMANAGER_H_
+
+#include "dart/common/AddonManager.h"
+#include "dart/common/AddonManagerJoiner.h"
+#include "dart/common/Virtual.h"
+
+namespace dart {
+namespace common {
+
+/// Declaration of the variadic template
+template <class... OtherSpecAddons>
+class SpecializedAddonManager { };
+
+//==============================================================================
+/// SpecializedAddonManager allows classes that inherit AddonManager to have
+/// constant-time access to a specific type of Addon
+template <class SpecAddon>
+class SpecializedAddonManager<SpecAddon> : public virtual AddonManager
+{
+public:
+
+  /// Default Constructor
+  SpecializedAddonManager();
+
+  /// Check if this AddonManager currently has a certain type of Addon
+  template <class T>
+  bool has() const;
+
+  /// Get a certain type of Addon from this AddonManager
+  template <class T>
+  T* get();
+
+  /// Get a certain type of Addon from this AddonManager
+  template <class T>
+  const T* get() const;
+
+  /// Make a clone of the addon and place the clone into this AddonManager. If
+  /// an Addon of the same type already exists in this AddonManager, the
+  /// existing Addon will be destroyed.
+  template <class T>
+  void set(const T* addon);
+
+  /// Use move semantics to place an addon into this AddonManager. If an Addon
+  /// of the same type already exists in this AddonManager, the existing Addon
+  /// will be destroyed.
+  template <class T>
+  void set(std::unique_ptr<T>&& addon);
+
+  /// Construct an Addon inside of this AddonManager
+  template <class T, typename ...Args>
+  T* create(Args&&... args);
+
+  /// Remove an Addon from this AddonManager
+  template <class T>
+  void erase();
+
+  /// Remove an Addon from this AddonManager, but return its unique_ptr instead
+  /// of letting it be deleted. This allows you to safely use move semantics to
+  /// transfer an Addon between two AddonManagers.
+  template <class T>
+  std::unique_ptr<T> release();
+
+  /// Check if this Manager is specialized for a specific type of Addon
+  template <class T>
+  static constexpr bool isSpecializedFor();
+
+protected:
+
+  /// Redirect to AddonManager::has()
+  template <class T>
+  bool _has(type<T>) const;
```

Comment:
Well, we don't have any documented convention for code styles yet, but I thought unofficially we prefer not to use leading underscores in naming. I would prefer to rename these method names just not to use the underscores. It wouldn't cause confliction with the public version functions since they have different parameters. Or we can consider using other names such as `hasImpl()`, but my preference would be the names just removed the underscores.